src/edit_mode/vi/mod.rs

```diff
@@ -264,4 +264,84 @@ mod test {
 
         assert_eq!(result, ReedlineEvent::None);
     }
+
+    #[allow(non_snake_case)]
+    mod integration {
+        use super::*;
+        use crate::{Reedline, ReedlineEvent};
+        use pretty_assertions::assert_eq;
+        use rstest::rstest;
+
+        fn unwrap_edits(event: ReedlineEvent) -> Vec<EditCommand> {
+            match event {
+                ReedlineEvent::Edit(c) => c,
+                // It's not clear what the difference is between these, but vi mode often returns Multiple
+                ReedlineEvent::Multiple(events) => {
+                    events.into_iter().flat_map(unwrap_edits).collect()
+                }
+                other => panic!("unexpected event {other:#?}"),
+            }
+        }
+
+        #[rstest]
+        #[case::edit_dw_P(&["^", "dw", "P"])]
+        #[case::edit_db_P(&["db", "P"])]
+        #[case::edit_dW_P(&["^", "dW", "P"])]
+        #[case::edit_dB_P(&["dB", "P"])]
+        #[case::edit_3dw_P(&["^", "3dw", "P"])]
+        #[case::edit_d3w_P(&["^", "d3w", "P"])]
+        #[case::edit_dtz_P(&["^", "dtz", "P"])]
+        #[case::edit_dTz_P(&["dTz", "P"])]
+        #[case::edit_dfz_P(&["^", "dfz", "P"])]
+        #[case::edit_dFz_P(&["dFz", "P"])]
+        #[case::edit_dw_dw_P(&[
+            // duplicate the first word, because dw dw P should drop the first word
+            "^", "dw", "P", "P", "Fj",
+            // run the actual test, and it should put us back where we started.
+            "dw", "dw", "P"])]
```

Comment:
This rstest is me trying to implement something like https://matklad.github.io/2021/05/31/how-to-test.html#Data-Driven-Testing

I tried to avoid writing test cases in terms of internal data structures like EditCommand because they tend not to be refactor-proof. Vim commands are already text based, so they feel like a natural base for a testing microformat/DSL. Might be worth making an equivalent DSL for Emacs using the `C-w` notation from the Emacs/bash docs. Probably something for another PR though.

I think this particular example might be stretching the usefulness of the interface I built a bit though. A better way to write this test would probably be on top of a `check_commands(before, commands, after)` helper, probably using $0/$1 to encode the cursor position, as the rust-analyzer tests do. The rest of the `sum_to_zero` test cases could then be rebased on top of such an abstraction.

I'm not sure when I will have another opportunity for recreational hacking. It's been wiping me out recently. I am planning to take some time off in October, so I will have another stab at it then if I haven't finished it before.